java-backend/src/main/java/org/kframework/backend/java/symbolic/InitializeRewriter.java

```diff
@@ -354,9 +358,7 @@ public ProcessProofRules invoke(GlobalContext rewritingContext, long initCounter
 
             private Term evaluate(Term term, ConjunctiveFormula constraint, TermContext context) {
                 context.setTopConstraint(constraint);
-                Term newTerm = term.fullSubstituteAndEvaluate(context);
-                context.setTopConstraint(null);
-                return newTerm;
+                return term.fullSubstituteAndEvaluate(context);
```

Comment:
could you please explain why `context.setTopConstraint(null);` is not needed here?

Reply:
@daejunpark I just reverted Andrei's commit that was setting `context.setTopConstraint(null);` in certain plaaces, his previous hotfix solution to the issue refererred in this PR description. Since this PR, that hotfix is no longer needed. Please look at commit history.

Reply:
As Andrei once proposed, Constraint should be separated from context and passed as argument to every method that needs it. Then things would be much clearer. But that would be a very big change that is not worth at this point when java-backend is in maintenance mode.

Reply:
if you are fully sure that the deleted line has no effect, then ok.  i'm still suspicious though. but we will find the problem if any later.

Reply:
Constraint is reset at the beginning of each rule evaluation:
https://github.com/kframework/k/blob/017c591d4b20802238afe431154659f232c8cca8/java-backend/src/main/java/org/kframework/backend/java/symbolic/InitializeRewriter.java#L350

So there's no need to reset it at the end. Code with no effect complicates understanding and should be removed. And we want to evaluate both LHS and RHS in the presence of same constraint. Before this PR only LHS was evaluated in the presence of constraint, judging by the code diff. So it was wrong.